lib/Model/NameBuilder.cpp

```diff
@@ -64,6 +66,90 @@ static size_t finalDigitCount(llvm::StringRef Name) {
   return Result;
 }
 
+static llvm::StringRef dropDeduplicatedWrapperSuffix(llvm::StringRef Name) {
+  if (size_t DigitCount = finalDigitCount(Name))
+    if (Name.substr(0, Name.size() - DigitCount).ends_with("_wrapper_"))
+      return Name.substr(0, Name.size() - DigitCount - 9);
+
+  return Name;
+}
+
+static std::vector<std::string> loadHelperFile(llvm::StringRef Path) {
+  auto MaybePath = revng::ResourceFinder.findFile(Path);
+  if (!MaybePath.has_value()) {
+    std::string Error = "Helper list is missing: " + Path.str();
+    revng_abort(Error.c_str());
+  }
+
+  auto Parsed = fromFileOrSTDIN<std::vector<std::string>>(MaybePath.value());
+  if (!Parsed) {
+    std::string Error = "Unable to deserialize the helper list: " + Path.str();
+    revng_abort(Error.c_str());
+  }
+
+  return std::move(*Parsed);
+}
+
+static std::string tcgHelperPath(model::Architecture::Values Architecture) {
+  return "share/revng/helper-lists/libtinycode-"
+         + model::Architecture::getQEMUName(Architecture).str() + ".bc.txt";
+}
+static std::string
+earlyLinkedHelperPath(model::Architecture::Values Architecture) {
+  return "share/revng/helper-lists/early-linked-"
+         + model::Architecture::getQEMUName(Architecture).str() + ".ll.txt";
+}
```

Comment:
Can we have a single file that contains names to be banned per-architecture?

To be frank. we could also consider to have a single file for all architectures: remember that eventually we'll support multi-architecture projects.

It'd made things simpler.

Also let's avoid mentioning TCG and QEMU, unless strictly necessary, the goal is to make rev.ng agnostic to those.

Reply:
Can you attach a zip with all the list of names so I can better evaluate?

My feeling is that we shouldn't try to be extra smart here.
I think a more reasonable approach to reduce the number of (likely) collisions would be to prefix all these functions with `revng_helper_` or something like that (not today, though).

Let's keep things simple.

Reply:
[helper-lists.zip](https://github.com/user-attachments/files/18981409/helper-lists.zip)

Reply:
1. If it's a `.txt` file don't add the leading `- `. It's not a YAML file.
2. There are `3725` distinct functions across them all.
   If we split per-architecture we get instead:

    ```
    x86_64: 1585
    s390x: 970
    mipsel: 1652
    mips: 1730
    i386: 1593
    mips: 1730
    arm: 1507
    aarch64: 1534
    ```

   The difference is 2x/3x.
   I'd prefer to a single list to keep things simple and avoid coupling validity of a name to (a set of) architectures.

   As mentioned before, I think the proper solution to reduce collisions is to devise a way to prefix these all.

Reply:
1. Maybe I'm just weird, but these leading `- `s make it easier to read! To the point of me considering to actually turn this into a yaml file. But yeah, that's definitely not a good idea - he reason to suggest we do any complex parsing dropping them it is.

2. I still think that the fewer names we ban the better, even at cost of a bit of extra complexity. But, yeah, considering we're going to make these names even uglier sooner or later, who cares. Let's merge them.

Reply:
Done!